### PR TITLE
Revised module resolution

### PR DESCRIPTION
**Pull request by vladima**
This PR consolidates all module resolution logic in one place so there is only one entry point for everything that needs to resolve module names. Currently it is function `getDefaultModuleNameResolver` function that accepts `CompilerOptions` and returns back an instance of `ModuleResolved` which is the alias for `export type ModuleNameResolver = (moduleName: string, containingFile: string, options: CompilerOptions, host: ModuleResolutionHost) => ResolvedModule;`. Default module resolver is used in several places:
- `preProcessFile` that is called from VS to determine dependencies of the file
- `createProgram` - at this point default module resolver will be requested if host wants to process with default module name resolution rules. NOTE: host can override built-in rules by implementing `resolveModuleNames?(moduleNames: string[], containingFile: string): string[];` function that accepts a list of imported names in the file and returns a list of resolved file names in the same order.
- `resolveModuleNames` implementation in `LSHost` for 'tsserver'.

Now `CompilerHost / LanguageServiceHost` can implement `resolveModuleNames` and be completely in control of how module names are resolved. Alternatively they might want to use built-in rules - in this case they should implement `getModuleResolutionHost`, return value of this function will be used by default module resolver to ask file system related questions.

### CODE REVIEW

**Comment by vladima** [comment]
In theory existing implementation of `LanguageServiceHost` interface should just work because of [this](https://github.com/Microsoft/TypeScript/pull/4154/files#diff-233e1126c0abc811c4098757f9e4516eR2589). Pinging @basarat so he can be aware of this change.

**Comment by basarat** [comment]
Excited :rose: !

**Comment by vladima** [comment]
adding @dbaeumer, @derekcicerone

File: src/harness/projectsRunner.ts [hunk 1]
```diff
@@ -129,6 +129,10 @@ class ProjectRunner extends RunnerBase {
             getSourceFileText: (fileName: string) => string,
             writeFile: (fileName: string, data: string, writeByteOrderMark: boolean) => void): CompileProjectFilesResult {
 
+            let moduleResolutionHost: ts.ModuleResolutionHost = {
+                fileExists: fileName => getSourceFile(fileName, ts.ScriptTarget.ES5) !== undefined,
```
Comment (mhegazy) [inline src/harness/projectsRunner.ts]: i would use Harness.IO.fileExists

File: src/compiler/core.ts [hunk 2]
```diff
@@ -709,7 +714,7 @@ namespace ts {
     /**
      *  List of supported extensions in order of file resolution precedence.
      */
-    export const supportedExtensions = [".tsx", ".ts", ".d.ts"];
+    export const supportedExtensions = [".ts", ".d.ts", ".tsx"];
```
Comment (mhegazy) [inline src/compiler/core.ts]: .tsx should come before .d.ts

File: src/compiler/program.ts [hunk 3]
```diff
@@ -158,24 +222,57 @@ namespace ts {
         let start = new Date().getTime();
 
         host = host || createCompilerHost(options);
+        
+        // initialize resolveModuleNameWorker only if noResolve is false
+        let resolveModuleNamesWorker: (moduleNames: string[], containingFile: string) => string[];
+        if (!options.noResolve) {
+            resolveModuleNamesWorker = host.resolveModuleNames;
+            if (!resolveModuleNamesWorker) {
+                Debug.assert(host.getModuleResolutionHost !== undefined);
```
Comment (mhegazy) [inline src/compiler/program.ts]: As we discussed, i think this should be directly on the host. this way you will get a compile-time error.

Comment (mhegazy) [inline src/compiler/program.ts]: also this is a breaking change anyways.

File: src/compiler/program.ts [hunk 4]
```diff
@@ -158,24 +222,57 @@ namespace ts {
...
-        if (!skipDefaultLib) {
-            processRootFile(host.getDefaultLibFileName(options), /*isDefaultLib:*/ true);
+        
+        if (oldProgram) {
+            // check properties that can affect structure of the program or module resolution strategy
+            // if any of these properties has changed - structure cannot be reused
+            let oldOptions = oldProgram.getCompilerOptions();
+            if ((oldOptions.module !== options.module) || 
+                (oldOptions.noResolve !== options.noResolve) || 
+                (oldOptions.target !== options.target) ||
```
Comment (mhegazy) [inline src/compiler/program.ts]: also add `options.jsx`

File: src/services/services.ts [hunk 5]
```diff
@@ -1940,7 +1953,7 @@ namespace ts {
         let getCanonicalFileName = createGetCanonicalFileName(!!useCaseSensitiveFileNames);
 
         function getKeyFromCompilationSettings(settings: CompilerOptions): string {
-            return "_" + settings.target; //  + "|" + settings.propagateEnumConstantoString()
+            return "_" + settings.target + "|" + settings.module + "|" + settings.noResolve;
```
Comment (mhegazy) [inline src/services/services.ts]: and settings.jsx

File: src/harness/projectsRunner.ts (hunk 1 above)
Comment (vladima) [inline src/harness/projectsRunner.ts]: done

File: src/compiler/core.ts (hunk 2 above)
Comment (vladima) [inline src/compiler/core.ts]: done

File: src/compiler/program.ts (hunk 4 above)
Comment (vladima) [inline src/compiler/program.ts]: done

File: src/services/services.ts (hunk 5 above)
Comment (vladima) [inline src/services/services.ts]: done

**Comment by derekcicerone** [comment]
Neat, looks like this could be used for modules which need to be resolved across project boundaries perhaps.

File: src/compiler/program.ts [hunk 6]
```diff
@@ -25,6 +28,62 @@ namespace ts {
...
+        return resolveModuleName;
+    }
+    
+    export function resolveTripleslashReference(moduleName: string, containingFile: string): string {
+        let basePath = getDirectoryPath(containingFile);
+        let referencedFileName = isRootedDiskPath(moduleName) ? moduleName : combinePaths(basePath, moduleName);
+        return normalizePath(referencedFileName);
+    }
+
+    function resolveModuleName(moduleName: string, containingFile: string, compilerOptions: CompilerOptions, host: ModuleResolutionHost): ResolvedModule {
```
Comment (mhegazy) [inline src/compiler/program.ts]:
```ts
 function resolveModuleName(moduleName: string, containingFile: string, compilerOptions: CompilerOptions, host: ModuleResolutionHost): ResolvedModule {
    switch(compilerOptions.module) {
         case ModuleKind.CommonJS:
               resolveCommonJSModuleName(....);
....
    }
}   

```

Comment (vladima) [inline src/compiler/program.ts]: done

**Comment by mhegazy** [comment]
:+1:

**Comment by jbrantly** [comment]
Just wanted to comment that I recently added support for this to [ts-loader](https://github.com/jbrantly/ts-loader) for resolving modules the same way that webpack does (setting up aliases and such) and it's fantastic. This is a great addition to TypeScript! Good work.

**Comment by vladima** [comment]
Great to hear that!

**Pull request merged**